packages/fetchai/skills/tac_control_contract/behaviours.py

```diff
@@ -0,0 +1,271 @@
+from packages.fetchai.skills.tac_control_contract.game import Game, Phase
+from packages.fetchai.skills.tac_control_contract.parameters import Parameters
+
+CONTROLLER_DATAMODEL = DataModel(
+    "tac",
+    [Attribute("version", str, True, "Version number of the TAC Controller Agent.")],
+)
+
+logger = logging.getLogger("aea.tac_control_skill")
+
+
+class TACBehaviour(Behaviour):
+    """This class implements the TAC control behaviour."""
+
+    def __init__(self, **kwargs):
+        """Instantiate the behaviour."""
+        super().__init__(**kwargs)
+        self._oef_msg_id = 0
+        self._registered_desc = None  # type: Optional[Description]
+
+    def setup(self) -> None:
+        """
+        Implement the setup.
+
+        :return: None
+        """
+        parameters = cast(Parameters, self.context.parameters)
+        contract = cast(Contract, self.context.contracts.erc1155)
+        ledger_api = cast(EthereumApi, self.context.ledger_apis.apis.get("ethereum"))
+
+        #  Deploy the contract if there is no address in the parameters
+        if parameters.contract_address is None:
+            contract.set_instance(ledger_api)
+            transaction_message = contract.get_deploy_transaction(
+                deployer_address=self.context.agent_address,
+                ledger_api=ledger_api,
+                skill_callback_id=self.context.skill_id,
+            )
+
+            self.context.decision_maker_message_queue.put_nowait(transaction_message)
+        else:
+            self.context.logger.info("Setting the address of the deployed contract")
+            contract.set_instance_w_address(
+                ledger_api=ledger_api,
+                contract_address=str(parameters.contract_address),
+            )
+            contract.item_ids = parameters.good_ids
+
+    def act(self) -> None:
+        """
+        Implement the act.
+
+        :return: None
+        """
+        game = cast(Game, self.context.game)
+        parameters = cast(Parameters, self.context.parameters)
+        now = datetime.datetime.now()
+        # contract = cast(Contract, self.context.contracts.erc1155)
+
+        # if (
+        #     contract.is_deployed
+        #     and game.phase.value == Phase.PRE_GAME.value
+        #     and now < parameters.registration_start_time
+        # ):
+        #     self.context.logger.info("Creating the items.")
+        #     transaction_message = self._create_items()
+        #     self.context.decision_maker_message_queue.put_nowait(transaction_message)
+
+        if (
+            game.phase.value == Phase.PRE_GAME.value
+            and parameters.registration_start_time < now < parameters.start_time
+        ):
+
+            game.phase = Phase.GAME_REGISTRATION
+            self._register_tac()
+            self.context.logger.info(
+                "[{}]: TAC open for registration until: {}".format(
+                    self.context.agent_name, parameters.start_time
+                )
+            )
```

Comment:
here I'd add `self._create_items()`. You can create the items before you start the game. However, you need a way to check whether the items already exist on the contract (in case it is an existing one). I'd suggest we have a simple items lookup method on the contract.

Reply:
We can assume for now that the items are unique in the contract and then change things inside the smart contract. I suspect that we will need to introduce more functionality to the smart contract. Or maybe we will need to switch the map of stored items to -> `map(item_id, map(address, amount))`. This will make it easier to request the already created items that exist in the smart contract. But will make more difficult to return the balance of an address.

Reply:
So as we changed this now on the contract, which token is the "currency" and which set of tokens are the "tokens" needs to be tracked here in the skill rather than on the contract.